source/drivers/MicroBitFileSystem.cpp

```diff
@@ -0,0 +1,1278 @@
+  *
+  * @param filename A fully or partially qualified filename.
+  * @param directory The directory to search. If ommitted, the root directory will be used.
+  *
+  * @return A pointer to the DirectoryEntry for the given file, or NULL if no entry is found.
+  */
+DirectoryEntry* MicroBitFileSystem::getDirectoryEntry(char const * filename, const DirectoryEntry *directory)
+{
+    Directory *dir;
+    char const *file;
+    uint16_t block;
+    DirectoryEntry *dirent;
+
+    // Determine the filename from the (potentially) fully qualified filename.
+    file = filename + strlen(filename);
+    while (file >= filename && *file != '/')
+        file--;
+    file++;
+
+    // Obtain a handle on the directory to search.
+    if (directory == NULL)
+        directory = rootDirectory;
+
+    block = directory->first_block;
+    dir = (Directory *) getBlock(block);
+    dirent = &dir->entry[0];
+
+    // Iterate through the directory entries until we find our file, or run out of space.
+    while (1)
+    {
+        if ((uint32_t)(dirent + 1) > (uint32_t)dir + MBFS_BLOCK_SIZE)
+        {
+            block = getNextFileBlock(block);
+            if (block == MBFS_EOF)
+                return NULL;
+
+            dir = (Directory *)getBlock(block);
+            dirent = &dir->entry[0];
+        }
+
+        // Check for a valid match
+        if (dirent->flags & MBFS_DIRECTORY_ENTRY_VALID && strcmp(dirent->file_name, file) == 0)
+            return dirent;
+
+        // Move onto the next entry.
+        dirent++;
+    }
+
+    return NULL;
+}
+
+/**
+  * Determine the number of logical blocks required to hold the file table.
+  *
+  * @return The number of logical blocks required to hold the file table.
+  */
+uint16_t MicroBitFileSystem::calculateFileTableSize()
+{
+    uint16_t size = (fileSystemSize * 2) / MBFS_BLOCK_SIZE;
+    if ((fileSystemSize * 2) % MBFS_BLOCK_SIZE)
+        size++;
+
+    return size;
+}
+
+/**
+  * Retrieve a memory pointer for the start of the physical memory page containing the given block.
+  *
+  * @param block A valid block number.
+  *
+  * @return A pointer to the physical page in FLASH memory holding the given block.
+  */
+uint32_t *MicroBitFileSystem::getPage(uint16_t block)
+{
+    uint32_t address = (uint32_t) getBlock(block);
+    return (uint32_t *) (address - address % PAGE_SIZE);
+}
+
+/**
+  * Retrieve a memory pointer for the start of the given block.
+  *
+  * @param block A valid block number.
+  *
+  * @return A pointer to the FLASH memory associated with the given block.
+  */
+uint32_t *MicroBitFileSystem::getBlock(uint16_t block)
+{
+    return (uint32_t *)((uint32_t)fileSystemTable + block * MBFS_BLOCK_SIZE);
+}
+
+/**
+  * Retrieve the next block in a chain.
+  *
+  * @param block A valid block number.
+  *
+  * @return The block number of the next block in the file.
+  */
+uint16_t MicroBitFileSystem::getNextFileBlock(uint16_t block)
+{
+    return fileSystemTable[block];
```

Comment:
:-) nope. The file table refers to the next block in a chain, so fileSystemTable[block] refers to the next block in sequence... so we're good here. This is the same principle as FAT really, just a bit leaner and optimised for RAM and memory mapped FLASH... (perhaps we should call it "THIN"!)